src/wasm/wasm-validator.cpp

```diff
@@ -3461,6 +3461,25 @@ static void validateTags(Module& module, ValidationInfo& info) {
   }
 }
 
+static void validateTypes(Module& module, ValidationInfo& info) {
+  for (auto type : ModuleUtils::collectHeapTypes(module)) {
```

Comment:
I am somewhat worried about scanning the entire module for this validation - it will basically double validation times, I think.

We already validate GC instructions, which are most of the places GC types can appear. Perhaps we can add validation in the other places, like function params and locals?

Reply:
Yeah, maybe. I'd be worried about missing places and duplicating the logic from `ModuleUtils::collectHeapTypes` (which we've had trouble remembering to update in just a single place), but for better or worse maybe the fuzzer will help find cases where we miss adding validation.

Reply:
Yeah, definitely a tradeoff. Not sure what's best.

Can we at least not do the extra pass when GC is not enabled? That is, if GC is not enabled then I'd expect to get lots of other validation errors if the code contains GC types. And the new check added here seems to mostly be important to differentiate things in the context of GC being enabled anyhow (nominal/isorecursive differences).

Reply:
Sure, that seems reasonable for now.

Reply:
I pushed a better solution just now. PTAL.